collector/arp_linux.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !noarp
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type arpCollector struct {
+	count *prometheus.Desc
+}
+
+func init() {
+	Factories["arp"] = NewARPCollector
+}
+
+// NewARPCollector returns a new Collector exposing ARP stats.
+func NewARPCollector() (Collector, error) {
+	return &arpCollector{
+		count: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "arp", "count"),
+			"ARP entries by device",
+			[]string{"device"}, nil,
+		),
+	}, nil
+}
+
+func getArpEntries() (map[string]uint32, error) {
+	file, err := os.Open(procFilePath("net/arp"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	entries, err := parseArpEntries(file)
+	if err != nil {
+		return nil, err
+	}
+
+	return entries, nil
+}
+
+// TODO: This should get extracted to the github.com/prometheus/procfs package
+// to support more complete parsing of /proc/net/arp. Instead of adding
+// more fields to this function's return values it should get moved and
+// changed to support each field.
+func parseArpEntries(data io.Reader) (map[string]uint32, error) {
+	scanner := bufio.NewScanner(data)
+	entries := make(map[string]uint32)
+
+	for scanner.Scan() {
+		columns := strings.Fields(scanner.Text())
+
+		if len(columns) > 0 {
```

Comment:
Sorry for all the nitpicking. Idiomatic go tries to reduce the amount of nesting as much as possible, so `if len(columns) == 0 { continue }` would be more common here. Though, I think we should actually return an error here, such line would be completely unexpected and tell us about either a broken or changed ARP file. So I suggest:

```go
if len(columns) < 6 {
    return nil, fmt.Errorf("unexpected ARP table file")
}

// ...
```